pkg/trillianclient/trillian_client.go

```diff
@@ -75,18 +185,170 @@ func unmarshalLogRoot(logRoot []byte) (types.LogRootV1, error) {
+// others wait on the condition variable until initialization completes. This
+// avoids head-of-line blocking while keeping state updates atomic.
+func (t *TrillianClient) ensureStarted(ctx context.Context) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.started {
+		return t.startErr
+	}
+
+	// Perform one-time initialization while holding the lock for simplicity.
+	// This blocks other ensureStarted callers until initialization completes.
+	cctx := ctx
+	var cancel context.CancelFunc
+	if _, ok := ctx.Deadline(); !ok {
+		cctx, cancel = context.WithTimeout(ctx, t.config.InitLatestRootTimeout)
+	}
+	if cancel != nil {
+		defer cancel()
+	}
+	slr, err := t.client.GetLatestSignedLogRoot(cctx, &trillian.GetLatestSignedLogRootRequest{LogId: t.logID})
 	if err != nil {
-		return types.LogRootV1{}, err
+		t.startErr = err
+		return err
+	}
+	if slr == nil || slr.SignedLogRoot == nil {
+		err = fmt.Errorf("nil signed log root")
+		t.startErr = err
+		return err
+	}
+	r, uerr := unmarshalLogRoot(slr.SignedLogRoot.LogRoot)
+	if uerr != nil {
+		t.startErr = uerr
+		return uerr
+	}
+
+	t.v = client.NewLogVerifier(rfc6962.DefaultHasher)
+	t.lc = client.New(t.logID, t.client, t.v, r)
+	t.snapshot.Store(rootSnapshot{root: r, signed: slr.SignedLogRoot})
+	t.started = true
+	t.startErr = nil
+
+	// Start updater and track it for graceful shutdown.
+	t.wg.Add(1)
+	go func() {
+		defer t.wg.Done()
+		t.updater()
+	}()
+	return nil
+}
+
+// updater waits for root changes using the LogClient and notifies waiters.
+// It uses the parsed root from WaitForRootUpdate and synthesizes a minimal
+// SignedLogRoot (LogRoot bytes only) to avoid an extra network round trip
+// per advancement.
+func (t *TrillianClient) updater() {
+	// Create backoff for retry logic with reasonable defaults
+	bo := backoff.Backoff{
+		Min:    100 * time.Millisecond, // Start with 100ms
+		Max:    30 * time.Second,       // Cap at 30s
+		Factor: 2.0,                    // Double each time
+		Jitter: true,                   // Add randomization
+	}
+	for {
+		// Wrap the WaitForRootUpdate call with backoff retry
+		var nr *types.LogRootV1
+		err := bo.Retry(t.bgCtx, func() error {
+			select {
+			case <-t.stopCh:
+				return fmt.Errorf("client stopped")
+			default:
+			}
+
+			ctx, cancel := context.WithTimeout(t.bgCtx, t.config.UpdaterWaitTimeout)
+			defer cancel()
+
+			var waitErr error
+			nr, waitErr = t.lc.WaitForRootUpdate(ctx)
+			return waitErr
+		})
+		select {
+		case <-t.stopCh:
+			return
+		default:
+		}
+
+		if err != nil {
+			log.Logger.Debugw("trillian root update wait failed after retries", "treeID", t.logID, "err", err)
+			metricUpdaterErrors.WithLabelValues(fmt.Sprintf("%d", t.logID)).Inc()
+			// Reset backoff on persistent failure and continue to next iteration
+			bo.Reset()
+			continue
+		}
+
+		// Success - reset backoff for next potential failure
+		bo.Reset()
+
+		if nr == nil {
```

Comment:
How can `nr` be nil here? There should only be one thread running `updater()`, correct?

Reply:
based on the current implementation of `WaitForRootUpdate`, this can never happen. but in case that were to change, if we don't have a new root then there's nothing to update.